traffic_ops/traffic_ops_golang/logs/log.go

```diff
@@ -121,26 +122,51 @@ func getLastSeenCookie(r *http.Request) (time.Time, bool) {
 	return lastSeen, true
 }
 
-func getLog(tx *sql.Tx, days int, limit int) ([]tc.Log, error) {
-	rows, err := tx.Query(`
+const selectFromQuery = `
 SELECT l.id, l.level, l.message, u.username as user, l.ticketnum, l.last_updated
-FROM "log" as l JOIN tm_user as u ON l.tm_user = u.id
-WHERE l.last_updated > now() - ($1 || ' DAY')::INTERVAL
-ORDER BY l.last_updated DESC
-LIMIT $2
-`, days, limit)
+FROM "log" as l JOIN tm_user as u ON l.tm_user = u.id`
+
+func getLog(inf *api.APIInfo, days int, limit int) ([]tc.Log, uint64, error) {
+	var count = uint64(0)
+	if _, ok := inf.Params["orderby"]; !ok {
+		inf.Params["orderby"] = "last_updated"
+	}
+
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		"username":     dbhelpers.WhereColumnInfo{Column: "u.username", Checker: nil},
+		"last_updated": dbhelpers.WhereColumnInfo{Column: "l.last_updated", Checker: nil},
+	}
+	where, orderBy, pagination, queryValues, errs :=
+		dbhelpers.BuildWhereAndOrderByAndPagination(inf.Params, queryParamsToQueryCols)
+	if len(errs) > 0 {
+		return nil, 0, util.JoinErrs(errs)
+	}
+
+	timeInterval := fmt.Sprintf("l.last_updated > now() - INTERVAL '%v' DAY", days)
+	if _, ok := inf.Params["username"]; ok {
+		where = where + " AND " + timeInterval
+	} else {
+		where = "\nWHERE " + timeInterval
+	}
+	if pagination == "" {
+		pagination = pagination + fmt.Sprintf("\nLIMIT %v", limit)
+	}
+	query := selectFromQuery + where + orderBy + pagination
+
+	rows, err := inf.Tx.NamedQuery(query, queryValues)
 	if err != nil {
-		return nil, errors.New("querying logs: " + err.Error())
+		return nil, count, errors.New("querying logs: " + err.Error())
 	}
 	ls := []tc.Log{}
 	for rows.Next() {
 		l := tc.Log{}
 		if err = rows.Scan(&l.ID, &l.Level, &l.Message, &l.User, &l.TicketNum, &l.LastUpdated); err != nil {
-			return nil, errors.New("scanning logs: " + err.Error())
+			return nil, count, errors.New("scanning logs: " + err.Error())
 		}
+		count += 1
 		ls = append(ls, l)
 	}
-	return ls, nil
+	return ls, count, nil
```

Comment:
So regarding `summary`'s `count` property, the docs define it to mean:

> _"`count` contains an unsigned integer that defines the total number of results that could possibly be returned given the non-pagination query parameters supplied by the client."_

... which means you can't just count the things you're returning to the client. You have to make a separate query that doesn't include any pagination. Typically, you can build this query by doing

```go
const countQuery = `
SELECT COUNT(*)
FROM logs
`

func get(args...) {
	// normal API stuff, getting the actual data etc.

	// this 'where' is the one that comes from the dbhelpers function from above
	query = countQuery + where
	var count uint64
	ns, err := inf.Tx.PrepareNamed(query)
	if err != nil {
		// handle the error
	}
	// 'queryValues' are, again, as returned by the dbhelpers function
	err = inf.Tx.NamedStmt(ns).QueryRow(queryValues).Scan(&count)
	if err != nil {
		// handle the error
	}
	// 'count' now holds the appropriate log count
	// do whatever else you need to do, write the response etc.
}
```


... but if that sounds like too much work to you, then don't worry about it, you don't need to add a `summary` to the endpoint, that's a totally optional extension for API endpoints.

Reply:
Given this; https://github.com/apache/trafficcontrol/issues/5451#issuecomment-764983956, above suggested changed doesn't make sense.

Reply:
that comment was about a possible alternative to this, but apparently there's enough value in it that we're doing it, so what I said on #5451 no longer matters.